tools/wptrunner/wptrunner/executors/runner.js

```diff
@@ -1,42 +1,44 @@
 document.title = '%(title)s';
 
-window.addEventListener(
-  "message",
-  function(event) {
-    window.message_queue.push(event);
-    window.process_next_event();
-  },
-  false
-);
-
-
-window.process_next_event = function() {
-  /* This function handles the next testdriver event. The presence of
-     window.testdriver_callback is used as a switch; when that function
-     is present we are able to handle the next event and when is is not
-     present we must wait. Therefore to drive the event processing, this
-     function must be called in two circumstances:
-       * Every time there is a new event that we may be able to handle
-       * Every time we set the callback function
-     This function unsets the callback, so no further testdriver actions
-     will be run until it is reset, which wptrunner does after it has
-     completed handling the current action.
-   */
-  if (!window.testdriver_callback) {
+var message_queue = [];
+var source = null;
+
+/**
+ * In preparation to execute a test, testdriver issues a number of messages
+ * which should be queued until execution time. During execution, testdriver
+ * consumes the queued messages by repeatedly "resuming" via `do_testharness`.
+ * The sequence of these events is variable, so the handler must account for
+ * cases where "resume" requests arrive before any messages are available (by
+ * caching the request's source) and where messages arrive before "resume"
+ * requests (by queuing the messages).
+ */
+window.addEventListener("message", function(event) {
+  if (!event.data) {
     return;
   }
-  var event = window.message_queue.shift();
-  if (!event) {
+
+  if (event.data.type === "testdriver-resume") {
+    source = event.source;
```

Comment:
I don't really understand the point of this; it seems like the identity of the test window is statically known here; it's `window.win`. Why do we need to keep updating it?